scripts/update-spec-links.sh

```diff
@@ -0,0 +1,718 @@
+   fi
+  fi
+done
+
+i=0
+for folder in "${dirs_in[@]}"; do
+  i=$(( i + 1 ))
+  [[ -n "$verbose" ]] && printf '[%d/%d]: Checking directory for files with links: %s ... ' "$i" "${#dirs_in[@]}" "$folder"
+  declare new_files=()
+  set -o noglob
+  IFS=$'\n' new_files+=( $( grep -E --files-with-matches --recursive --no-messages --binary-file=without-match "$link_rx" "$folder" | sort ) )
+  set +o noglob
+  if [[ "${#new_files[@]}" -ne '0' ]]; then
+    [[ -n "$verbose" ]] && printf 'Found %d file(s) with links:\n' "${#new_files[@]}"
+    j=0
+    for file in "${new_files[@]}"; do
+      j=$(( j + 1 ))
+      [[ -n "$verbose" ]] && printf '[%d/%d|%d/%d]:   %s' "$i" "${#dirs_in[@]}" "$j" "${#new_files[@]}" "$file"
+      if [[ "$file" =~ \.md$ ]]; then
+        files+=( "$file" )
+        [[ -n "$verbose" ]] && printf '\n'
+      else
+        [[ -n "$verbose" ]] && printf ' (ignored, not a markdown file)\n'
+      fi
+    done
+  else
+    if [[ -n "$verbose" ]]; then
+      printf 'None found.\n'
+    elif [[ -z "$quiet" ]]; then
+      printf 'Directory does not have any files with links: %s\n' "$folder"
+    fi
+  fi
+done
+
+if [[ "${#files[@]}" -eq '0' ]]; then
+  printf 'No files found to update.\n'
+  exit 0
+fi
+
+[[ -z "$quiet" ]] && printf 'Updating links in %d file(s).\n' "${#files[@]}"
+[[ -n "$verbose" ]] && printf '  %s\n' "${files[@]}"
+
+########################################################################################################################
+###########################################  Identify proto files involved.  ###########################################
+########################################################################################################################
+
+[[ -z "$quiet" ]] && printf 'Identifying proto files linked to from %d markdown files.\n' "${#files[@]}"
+
+declare protos_linked=()
+i=0
+for file in "${files[@]}"; do
+  i=$(( i + 1 ))
+  [[ -n "$verbose" ]] && printf '[%d/%d]: Checking %s ... ' "$i" "${#files[@]}" "$file"
+  declare new_proto_files=()
+  set -o noglob
+  IFS=$'\n' new_proto_files+=( $( grep -E "$link_rx" "$file" | sed 's|^.*/proto/|proto/|; s/#.*$//;' | sort -u ) )
+  set +o noglob
+  if [[ "${#new_proto_files[@]}" -ne '0' ]]; then
+    if [[ -n "$verbose" ]]; then
+      printf 'Found %d:\n' "${#new_proto_files[@]}"
+      j=0
+      for new_file in "${new_proto_files[@]}"; do
+        j=$(( j + 1 ))
+        printf '[%d/%d|%d/%d]:   %s\n' "$i" "${#files[@]}" "$j" "${#new_proto_files[@]}" "$new_file"
+      done
+    fi
+    protos_linked+=( "${new_proto_files[@]}" )
+  else
+    [[ -n "$verbose" ]] && printf 'None found.\n'
+  fi
+done
+
+[[ -n "$verbose" ]] && printf 'All linked protos (with dups) (%d):\n' "${#protos_linked[@]}" && printf '  %s\n' "${protos_linked[@]}"
+
+# Deduplicate entries linked from multiple files.
+declare protos=()
+set -o noglob
+IFS=$'\n' protos+=( $( printf '%s\n' "${protos_linked[@]}" | sort -u ) )
```

Comment:
Again, `mapfile` is not something we can reliably expect environments to have.
Similarly, not all systems have a `read` utility that recognizes the `-a` flag.

Reply:
Continuing this conversation here: https://github.com/provenance-io/provenance/pull/2068#discussion_r1669060594 since you're just saying the same things.